doc/platform-optimization/README.md

```diff
@@ -0,0 +1,188 @@
+# Kubernetes Platform Optimization with Resource Management
+
+## Introduction
+
+This document provides an example how to manage which CPUs and
+memories (NUMA nodes) are allowed to be used by containers on a
+Kubernetes node.
+
+Managing CPUs and memories enables improving AI container performance
+and maintaining predictable response times even under heavy
+load. Reasons for performance improvements include the following.
+
+- Better cache hit ratios in all cache levels.
+- Fewer remote memory accesses.
+- Fewer processes and threads per CPU in the whole system.
+- Disabling CPU hyperthreading on containers that run faster when the
+  other CPU thread is idle.
+
+More predictable response times are possible by using dedicated CPUs
+for containers and sets of containers. This ensures that critical
+containers will always have enough compute resources, and that
+resource hungry containers will not be able to hurt all processes in
+the system.
+
+## NRI Plugins
+
+[NRI plugins](https://github.com/containers/nri-plugins) connect to
+the container runtime running on a Kubernetes node. Containerd and
+CRI-O runtimes support NRI plugins.
+
+The NRI plugins project includes two resource policies, balloons and
+topology-aware. They manage allowed CPUs and memories (cpuset.cpus and
+cpuset.mems) of all Kubernetes containers created and running on the
+node.
+
+In this example, we use the balloons policy because it can be tuned
+for certain applications (like RAG pipelines) using even node-specific
+parameters for each container in applications. The topology-aware
+policy, on the other hand, needs no configuration and does CPU
+assignment automatically based on resource requests in containers and
+underlying hardware topology.
+
+## Install
+
+Warning: installing and reconfiguring the balloons policy in can
```

Comment:
s/in/

Also how can one tell if there are containers running that are relying on some static allocation?

Reply:
It's impossible to automatically detect all containers that need static allowed CPUs, though there are some signs (for instance application setting CPU affinity of its own processes/threads) that suggest so.

Currently there is no generally agreed way to annotate a container with a "I can't handle my allowed CPUs changing". Perhaps this kind of annotation would enable runtime to restart/recreate containers whose CPUs need to be changed.